### PR TITLE
fix: show parsed platform-address transitions

### LINKED ISSUE #573: Can't deserialize platform address txs
When raw txs for platform address related STs are copied from platform explorer, they can't be deserialized by DET (v1.0-dev). I don't see any messages in det.log about this.

Example:
https://testnet.platform-explorer.com/transaction/B88F54B03619D99482FE98EA8DC9C288C3F4BB5FBF93131E703B5459F50BB78D

```DAABAEZJ8lqSWSBK4JRgM6A4o7EoAM/IDPwAD0JAAQA2I7Ax2quQXVnyAKnezuNPE75q7vwAD0JAAQAAAAEAQR/vseNQooUa8QVJbujMgP22M8EzL3C9AAEibx6iMtmWHA1ou89lGCCtHgJmKcieCtdhZMhqcnU9O+fc4Y575ryF```

Image

### PR DESCRIPTION
**Pull request by thepastaclaw**
# PR Description

## Summary

- show a debug JSON fallback when a parsed state transition cannot be
  serialized as JSON
- retain the parsed `StateTransition` separately so fallback-rendered
  transitions can still be broadcast
- add coverage for the issue #573 Platform-address transition payload

Fixes dashpay/dash-evo-tool#573.

## Validation

- `cargo fmt --all`
- focused unit test for
  `platform_address_transition_with_non_string_map_keys_uses_debug_fallback`
- `cargo clippy --lib -- -D warnings`
- pre-PR code review gate: ship

## Release Notes

* **Bug Fixes**
  * Improved reliability of “Broadcast Transition to Platform” by using the already-parsed transition data instead of re-parsing from the JSON display.
  * Enhanced state transition serialization and error handling: when serialization fails, the UI now shows a clear warning plus a debug-style fallback rather than failing silently.
  * Added coverage for the edge case where JSON serialization encounters non-string map keys.

### CODE REVIEW

**Comment by thepastaclaw** [comment]
@coderabbitai review

**Comment by thepastaclaw** [comment]
✅ Final review complete — no blockers (commit 07680086) · triage: critical · Phase 2 only (queue backlog)

File: src/ui/tools/transition_visualizer_screen.rs [hunk 1]
```diff
@@ -18,10 +18,37 @@ use dash_sdk::dpp::state_transition::StateTransition;
...
+    match serde_json::to_string_pretty(state_transition) {
+        Ok(json) => Ok(json),
+        Err(e) => {
+            let error = e.to_string();
+            let debug_output = format!("{:#?}", state_transition);
+            tracing::warn!(
+                error = %error,
+                state_transition = %debug_output,
+                "parsed state transition but failed to serialize it to JSON; using debug fallback"
+            );
```
[the remark this answers is not in the document]
Comment (thepastaclaw) [reply src/ui/tools/transition_visualizer_screen.rs:49]: Fixed in 9f9952b9: WARN now logs only the serialization error/context; the full debug transition is emitted at DEBUG instead.

File: src/ui/tools/transition_visualizer_screen.rs [hunk 2]
```diff
@@ -228,23 +261,26 @@ impl TransitionVisualizerScreen {
...
-                            && let Ok(state_transition) = serde_json::from_str(json)
-                        {
                             app_action = AppAction::BackendTask(
                                 BackendTask::BroadcastStateTransition(state_transition),
                             );
+                        } else {
+                            self.parse_error = Some((
+                                "Parsed transition is unavailable for broadcast".to_string(),
+                                Instant::now(),
+                            ));
```
[the remark this answers is not in the document]
Comment (thepastaclaw) [reply src/ui/tools/transition_visualizer_screen.rs]: Fixed in 9f9952b9: the broadcast-unavailable path now uses MessageBanner::set_global(..., MessageType::Error) instead of parse_error.

**Review by thepastaclaw — COMMENTED** [review]
## Code Review

The core fix — retaining the typed StateTransition so AddressFundsTransfer payloads can still be broadcast — is sound. Validated findings cluster around three areas: warn-level log noise on every keystroke, an unreachable else-branch from coupled Option<...> state, and a fragile test that asserts on serde_json's internal error wording. The debug fallback is embedded as a JSON string, which limits the readability gain advertised in the PR title.

*Reviewed commit: 9f9952b9*

Fresh dispatcher rerun for this queue item. Same-SHA automated reviews already existed, so I am posting this as a top-level review to avoid duplicating inline threads while still recording the fresh verification.

🟡 4 suggestion(s) | 💬 2 nitpick(s)

<details>
<summary>🟡 <b>suggestion:</b> warn-level log fires on every keystroke during typing</summary>

`src/ui/tools/transition_visualizer_screen.rs (lines 173-177)`

`parse_input()` runs whenever the input field changes (line 204–207), so this `tracing::warn!` will fire for every intermediate state where the user has typed bytes that decode but do not yet form a complete StateTransition. The result is warning-level log spam during normal interactive use, which makes real warnings harder to spot and adds avoidable work on the UI thread. Either downgrade to `debug`/`trace`, or only log when the user explicitly triggers parsing (e.g. on blur or via a button). The deserialize-failure case is the expected state during typing, not an exceptional condition.

<details>
<summary>💡 Suggested change</summary>

```suggestion
                    Err(e) => {
                        tracing::debug!(
                            error = %e,
                            "failed to deserialize state transition input"
                        );
                        self.parse_error =
                            Some((format!("Failed to parse: {}", e), Instant::now()));
                    }
```

</details>

</details>

<details>
<summary>🟡 <b>suggestion:</b> Coupled Option fields create an unreachable error branch</summary>

`src/ui/tools/transition_visualizer_screen.rs (lines 282-288)`

`parsed_json` and `parsed_state_transition` are always set together (lines 157–158) and cleared together (lines 121–122), and the broadcast button only renders when `parsed_json.is_some()` (line 244). The `else` arm here can therefore never execute in practice, and its banner text (`"Parsed transition is unavailable for broadcast"`) is jargon with no self-service action — it would not pass the project's user-facing message rules if it ever did fire. Collapse the two correlated fields into a single `Option<(StateTransition, String)>` (or a small `ParsedTransition { st, json }` struct). That makes the invariant type-enforced, removes this dead branch, and lets the broadcast path drop the `is_some()` check entirely.

</details>

<details>
<summary>🟡 <b>suggestion:</b> Debug fallback is embedded inside a JSON string and renders escaped</summary>

`src/ui/tools/transition_visualizer_screen.rs (lines 40-44)`

`serde_json::to_string_pretty` escapes newlines in string values, so the multi-line `{:#?}` output ends up as a single line containing literal `\n` sequences inside the `"debug"` field. For exactly the transitions that hit this fallback (the case the PR is meant to fix), the user sees an unreadable escaped blob instead of structured debug output — only the `warning` and `serialization_error` keys are useful. Consider either returning a plain-text rendering for the fallback (header + raw `{:#?}` body) so newlines are preserved in the `TextEdit::multiline`, or rendering the debug section in a separate non-JSON pane.

</details>

<details>
<summary>🟡 <b>suggestion:</b> Test does not cover the screen-level broadcast invariant introduced by this PR</summary>

`src/ui/tools/transition_visualizer_screen.rs (lines 566-581)`

The production change isn't just the JSON fallback helper — it's the combination of storing `parsed_state_transition` during parsing (line 158) and using that retained value in the broadcast button (line 267). The new test calls `state_transition_to_pretty_json()` directly, so it would still pass if a future refactor removed `self.parsed_state_transition = Some(state_transition)` or reverted the broadcast path to re-parsing `self.parsed_json`. A focused screen-level test (or kittest) should assert that feeding the issue #573 payload through the screen produces an `AppAction::BackendTask(BackendTask::BroadcastStateTransition(_))` even when the rendered pane is the debug fallback.

</details>

<details>
<summary>💬 <b>nitpick:</b> Assertion couples to serde_json's internal error wording</summary>

`src/ui/tools/transition_visualizer_screen.rs (lines 578-579)`

`assert!(json.contains("key must be a string"))` depends on the exact text of `serde_json::Error::Display` for non-string map keys, which is upstream's free-form message and can change across patch releases. The project guidance against depending on free-form error strings applies here in spirit. Assert on what we control instead: parse the JSON structurally and check that the `serialization_error` and `debug` keys are present and non-empty, and that `debug` contains a recognizable variant string from `StateTransition`'s `Debug` output. The other two asserts already demonstrate the fallback path was taken with the expected variant.

</details>

<details>
<summary>💬 <b>nitpick:</b> Helper returns Result<String, String> and the Err arm is effectively unreachable</summary>

`src/ui/tools/transition_visualizer_screen.rs (lines 25-53)`

The fallback object built at line 40–44 contains only `String` values, so `serde_json::to_string_pretty` will not fail in practice. The outer `Err` arm at the call site (line 168–170) is therefore dead, and the message it would surface (`"Failed to serialize state transition to JSON (...) and failed to build debug fallback (...)"`) leaks raw serde error text into the UI in violation of the project's user-facing message rules. Either make the helper infallible (return `String` and only log internally on the unlikely fallback failure), or — if a fallible signature is preferred — introduce a small `thiserror`-derived enum and surface a user-appropriate message at the call site with technical details attached via `BannerHandle::with_details(e)`.

</details>

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `src/ui/tools/transition_visualizer_screen.rs`:
- [SUGGESTION] lines 173-177: warn-level log fires on every keystroke during typing
  `parse_input()` runs whenever the input field changes (line 204–207), so this `tracing::warn!` will fire for every intermediate state where the user has typed bytes that decode but do not yet form a complete StateTransition. The result is warning-level log spam during normal interactive use, which makes real warnings harder to spot and adds avoidable work on the UI thread. Either downgrade to `debug`/`trace`, or only log when the user explicitly triggers parsing (e.g. on blur or via a button). The deserialize-failure case is the expected state during typing, not an exceptional condition.
- [SUGGESTION] lines 282-288: Coupled Option fields create an unreachable error branch
  `parsed_json` and `parsed_state_transition` are always set together (lines 157–158) and cleared together (lines 121–122), and the broadcast button only renders when `parsed_json.is_some()` (line 244). The `else` arm here can therefore never execute in practice, and its banner text (`"Parsed transition is unavailable for broadcast"`) is jargon with no self-service action — it would not pass the project's user-facing message rules if it ever did fire. Collapse the two correlated fields into a single `Option<(StateTransition, String)>` (or a small `ParsedTransition { st, json }` struct). That makes the invariant type-enforced, removes this dead branch, and lets the broadcast path drop the `is_some()` check entirely.
- [SUGGESTION] lines 40-44: Debug fallback is embedded inside a JSON string and renders escaped
  `serde_json::to_string_pretty` escapes newlines in string values, so the multi-line `{:#?}` output ends up as a single line containing literal `\n` sequences inside the `"debug"` field. For exactly the transitions that hit this fallback (the case the PR is meant to fix), the user sees an unreadable escaped blob instead of structured debug output — only the `warning` and `serialization_error` keys are useful. Consider either returning a plain-text rendering for the fallback (header + raw `{:#?}` body) so newlines are preserved in the `TextEdit::multiline`, or rendering the debug section in a separate non-JSON pane.
- [SUGGESTION] lines 566-581: Test does not cover the screen-level broadcast invariant introduced by this PR
  The production change isn't just the JSON fallback helper — it's the combination of storing `parsed_state_transition` during parsing (line 158) and using that retained value in the broadcast button (line 267). The new test calls `state_transition_to_pretty_json()` directly, so it would still pass if a future refactor removed `self.parsed_state_transition = Some(state_transition)` or reverted the broadcast path to re-parsing `self.parsed_json`. A focused screen-level test (or kittest) should assert that feeding the issue #573 payload through the screen produces an `AppAction::BackendTask(BackendTask::BroadcastStateTransition(_))` even when the rendered pane is the debug fallback.
```
</details>

**Review by thepastaclaw — COMMENTED** [review]
## Code Review

Narrowly scoped fix that correctly decouples broadcast from JSON serialization for transitions whose internal map keys aren't string-renderable. No correctness bugs found, but the change introduces a coupled-Option invariant with a dead error branch (and a non-compliant user-facing message), leaves contract-ID discovery dependent on the lossy rendered JSON, and adds a regression test that exercises the formatter rather than the retained-StateTransition invariant the fix actually relies on.

*Reviewed commit: 9f9952b9*

Fresh dispatcher gate run for the claimed queue item. `review_poster.py` detected an existing same-SHA review, so this records the fresh verification as a top-level review without duplicating inline threads.

🟡 3 suggestion(s) | 💬 2 nitpick(s)

<details>
<summary>🟡 <b>suggestion:</b> Combine `parsed_json` and `parsed_state_transition`; the `else` branch is unreachable and its message violates the user-message rules</summary>

`src/ui/tools/transition_visualizer_screen.rs (lines 264-289)`

`parsed_state_transition` is only ever assigned alongside `parsed_json` (lines 157-158) and cleared with it (121-122). The Broadcast button renders strictly inside `if let Some(ref json) = self.parsed_json` (244), so at click time `self.parsed_state_transition.clone()` is always `Some(_)` and the `else` arm at 282-288 is dead code. The two fields encode a single invariant the type system isn't enforcing — model them as one `Option<(String, StateTransition)>` (or a small named struct) so the invariant is structural and `parse_input` can't ever desync them. As a side effect, the unreachable error banner goes away — its wording ("Parsed transition is unavailable for broadcast") also fails the CLAUDE.md user-message rule of "what happened + what to do" (no concrete self-service action), so removing it is a net win.

</details>

<details>
<summary>🟡 <b>suggestion:</b> Contract-ID detection still keys off the rendered JSON, silently losing the "Fetch Contract" affordance on the fallback path</summary>

`src/ui/tools/transition_visualizer_screen.rs (lines 153-166)`

`state_transition_to_pretty_json` returns `Ok(_)` for both the real serialization and the debug-fallback wrapper, so `parsed_json` is always set when deserialization succeeds. But `extract_contract_ids` then walks that string. On the fallback path the string is `{warning, serialization_error, debug: "<format!('{:#?}', ...)>"}` — a single opaque blob — so any transition variant that both hits the fallback and carries `singleContract` bounds will silently lose its contract reference, and the user will never see the "Fetch contract" link even though deserialization fully succeeded. There is no real harm today (the platform-address transition this PR targets has no contract bound), but the data is now in two representations of different fidelity and behavior is keyed off the lossy one. Drive contract-ID discovery from the typed `StateTransition` (which is now retained on `self`) before rendering, instead of round-tripping through the JSON string.

</details>

<details>
<summary>🟡 <b>suggestion:</b> Test does not exercise the broadcast path or the retained-`StateTransition` invariant</summary>

`src/ui/tools/transition_visualizer_screen.rs (lines 566-581)`

The actual #573 fix is split across two behaviors: `parse_input()` now retains the typed transition in `parsed_state_transition` (157-158), and Broadcast sends that stored value rather than reparsing `parsed_json` (267-281). The new test only calls `state_transition_to_pretty_json()`, so a future refactor that drops `self.parsed_state_transition = Some(state_transition)` — or reverts Broadcast to `serde_json::from_str(json)` — would still pass. Add a screen-level test that drives `parse_input()` with the issue #573 payload and asserts both `parsed_json.is_some()` and `parsed_state_transition.is_some()` (the test module already has access to private fields).

</details>

<details>
<summary>💬 <b>nitpick:</b> Test asserts on serde/dpp-internal wording</summary>

`src/ui/tools/transition_visualizer_screen.rs (lines 578-580)`

`json.contains("key must be a string")` couples to serde_json's internal error string for non-string map keys, and `json.contains("AddressFundsTransfer")` couples to a dpp variant name. Either can shift without changing behavior this PR cares about. Prefer structural assertions: parse `json` back as `Value` and check that `warning`, `serialization_error`, and `debug` keys are present and non-empty (the warning literal is the only string this PR controls).

</details>

<details>
<summary>💬 <b>nitpick:</b> `Result<String, String>` is stringly-typed and the `Err` arm is effectively unreachable</summary>

`src/ui/tools/transition_visualizer_screen.rs (lines 25-53)`

CLAUDE.md error-handling guidance discourages stringified errors. In practice the outer `serde_json::to_string_pretty` over the `json!{...}` literal here has no failure mode (no NaN/Infinity floats, no non-string map keys), so the `map_err` arm is unreachable. Either drop the `Result` and return `String` infallibly (with `.expect()` on the fallback to assert the invariant), or surface a typed error (`Result<String, serde_json::Error>` or a small enum) so callers don't have to format and re-parse. Low impact since the only callsite consumes the message as a `parse_error` display, but this is a fresh helper and worth getting right.

</details>

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `src/ui/tools/transition_visualizer_screen.rs`:
- [SUGGESTION] lines 264-289: Combine `parsed_json` and `parsed_state_transition`; the `else` branch is unreachable and its message violates the user-message rules
  `parsed_state_transition` is only ever assigned alongside `parsed_json` (lines 157-158) and cleared with it (121-122). The Broadcast button renders strictly inside `if let Some(ref json) = self.parsed_json` (244), so at click time `self.parsed_state_transition.clone()` is always `Some(_)` and the `else` arm at 282-288 is dead code. The two fields encode a single invariant the type system isn't enforcing — model them as one `Option<(String, StateTransition)>` (or a small named struct) so the invariant is structural and `parse_input` can't ever desync them. As a side effect, the unreachable error banner goes away — its wording ("Parsed transition is unavailable for broadcast") also fails the CLAUDE.md user-message rule of "what happened + what to do" (no concrete self-service action), so removing it is a net win.
- [SUGGESTION] lines 153-166: Contract-ID detection still keys off the rendered JSON, silently losing the "Fetch Contract" affordance on the fallback path
  `state_transition_to_pretty_json` returns `Ok(_)` for both the real serialization and the debug-fallback wrapper, so `parsed_json` is always set when deserialization succeeds. But `extract_contract_ids` then walks that string. On the fallback path the string is `{warning, serialization_error, debug: "<format!('{:#?}', ...)>"}` — a single opaque blob — so any transition variant that both hits the fallback and carries `singleContract` bounds will silently lose its contract reference, and the user will never see the "Fetch contract" link even though deserialization fully succeeded. There is no real harm today (the platform-address transition this PR targets has no contract bound), but the data is now in two representations of different fidelity and behavior is keyed off the lossy one. Drive contract-ID discovery from the typed `StateTransition` (which is now retained on `self`) before rendering, instead of round-tripping through the JSON string.
- [SUGGESTION] lines 566-581: Test does not exercise the broadcast path or the retained-`StateTransition` invariant
  The actual #573 fix is split across two behaviors: `parse_input()` now retains the typed transition in `parsed_state_transition` (157-158), and Broadcast sends that stored value rather than reparsing `parsed_json` (267-281). The new test only calls `state_transition_to_pretty_json()`, so a future refactor that drops `self.parsed_state_transition = Some(state_transition)` — or reverts Broadcast to `serde_json::from_str(json)` — would still pass. Add a screen-level test that drives `parse_input()` with the issue #573 payload and asserts both `parsed_json.is_some()` and `parsed_state_transition.is_some()` (the test module already has access to private fields).
```
</details>

**Comment by thepastaclaw** [comment]
@coderabbitai review

**Comment by thepastaclaw** [comment]
@coderabbitai review

**Comment by thepastaclaw** [comment]
@coderabbitai review

**Comment by thepastaclaw** [comment]
@coderabbitai review

**Comment by thepastaclaw** [comment]
@coderabbitai review

**Comment by thepastaclaw** [comment]
@coderabbitai review

**Comment by thepastaclaw** [comment]
@coderabbitai review

**Comment by thepastaclaw** [comment]
@coderabbitai review

**Comment by thepastaclaw** [comment]
@coderabbitai review

**Comment by thepastaclaw** [comment]
@coderabbitai review

**Review by thepastaclaw — COMMENTED** [review]
## Final validation — Phase 2 only (queue backlog)

The retained typed transition removes broadcasting's dependency on the rendered JSON, and the fallback preserves readable debug output. Both Phase-2 suggestions are supported by the source: the new fixture unnecessarily mutates process-global configuration, and the pinned address-map serializers leave the fallback path untested. No blocking defects were verified.

Source: reviewer 1: `gpt-6-astra` (agent: `phase2-reviewer`, role: `general`); reviewer 2: `gpt-6-astra` (agent: `phase2-reviewer`, role: `rust-quality`); final verifier: `gpt-6-astra` (agent: `astra-verifier`, role: `final-verifier`)

### Review provenance
- Triage: `critical` by `gpt-6-astra` (effort low) — Although contained to one UI screen, the change alters the broadcast path for state transitions, including fund transfers, so a mismatch between displayed and retained transition data could have financial consequences.
- Phase 1 reviewers: **not run (skipped for throughput: 63 PRs queued, above the 10 limit)**
- Fresh verifier: `gpt-6-astra` — final-verifier; agent `astra-verifier`
- Phase 2 reviewers: `gpt-6-astra` — general (completed, effort xhigh); agent `phase2-reviewer`, `gpt-6-astra` — rust-quality (completed, effort xhigh); agent `phase2-reviewer`

🟡 2 suggestion(s)

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `src/ui/tools/transition_visualizer_screen.rs`:
- [SUGGESTION] src/ui/tools/transition_visualizer_screen.rs:603-620: Remove redundant process-global setup from the test fixture
  The module-local Once only synchronizes calls to this initializer; it does not prevent unrelated tests or native threads from reading the environment while these unsafe set_var calls run. The initializer also calls copy_env_file_if_not_exists(), which can write outside the temporary test directory. Neither operation is needed: make_test_screen creates a bundled .env in its temporary directory, and AppContext::new loads it through Config::load_from. Remove ensure_test_env and its call, retaining the temporary-directory configuration or reusing the existing context test fixture. Config::load_from's own pre-existing environment mutation can remain a separate concern.
- [SUGGESTION] src/ui/tools/transition_visualizer_screen.rs:683-694: Exercise the serialization-failure fallback deterministically
  Both new tests use ISSUE_573_BASE64. On the pinned Platform revision, AddressFundsTransfer's input and output maps use serializers that emit arrays, so this fixture follows the successful JSON path. Selecting the assertions with the same serializer therefore leaves the new failure branch and its newline preservation untested; replacing that branch with a panic would not be caught by these tests. Keep the historical fixture, but add a deterministic failure-path test using a demonstrably failing transition or a small injectable serialization boundary. Explicitly assert the serialization failure and readable multiline fallback output.
```
</details>

File: src/ui/tools/transition_visualizer_screen.rs [hunk 3]
```diff
@@ -539,3 +580,161 @@ impl ScreenLike for TransitionVisualizerScreen {
...
+                std::env::set_var("MAINNET_core_host", "127.0.0.1");
+                std::env::set_var("MAINNET_core_rpc_port", "9998");
+                std::env::set_var("MAINNET_core_rpc_user", "dashrpc");
+                std::env::set_var("MAINNET_core_rpc_password", "password");
+
+                std::env::set_var("LOCAL_dapi_addresses", "http://127.0.0.1:2443");
+                std::env::set_var("LOCAL_core_host", "127.0.0.1");
+                std::env::set_var("LOCAL_core_rpc_port", "20302");
+                std::env::set_var("LOCAL_core_rpc_user", "dashmate");
+                std::env::set_var("LOCAL_core_rpc_password", "password");
```
Comment (thepastaclaw) [inline src/ui/tools/transition_visualizer_screen.rs:620]: **🟡 Suggestion: Remove redundant process-global setup from the test fixture**

The module-local Once only synchronizes calls to this initializer; it does not prevent unrelated tests or native threads from reading the environment while these unsafe set_var calls run. The initializer also calls copy_env_file_if_not_exists(), which can write outside the temporary test directory. Neither operation is needed: make_test_screen creates a bundled .env in its temporary directory, and AppContext::new loads it through Config::load_from. Remove ensure_test_env and its call, retaining the temporary-directory configuration or reusing the existing context test fixture. Config::load_from's own pre-existing environment mutation can remain a separate concern.

<sub>source: ['claude']</sub>

File: src/ui/tools/transition_visualizer_screen.rs [hunk 4]
```diff
@@ -539,3 +580,161 @@ impl ScreenLike for TransitionVisualizerScreen {
...
+            // Prefer the JSON path when the current pin can serialize it.
+            assert!(
+                rendered.trim_start().starts_with('{'),
+                "successful JSON serialization should yield pretty JSON"
+            );
+            assert!(
+                serde_json::from_str::<Value>(&rendered).is_ok(),
+                "JSON path must produce parseable JSON"
+            );
+        } else {
```
Comment (thepastaclaw) [inline src/ui/tools/transition_visualizer_screen.rs:694]: **🟡 Suggestion: Exercise the serialization-failure fallback deterministically**

Both new tests use ISSUE_573_BASE64. On the pinned Platform revision, AddressFundsTransfer's input and output maps use serializers that emit arrays, so this fixture follows the successful JSON path. Selecting the assertions with the same serializer therefore leaves the new failure branch and its newline preservation untested; replacing that branch with a panic would not be caught by these tests. Keep the historical fixture, but add a deterministic failure-path test using a demonstrably failing transition or a small injectable serialization boundary. Explicitly assert the serialization failure and readable multiline fallback output.

<sub>source: ['claude']</sub>

**Comment by lklimek** [comment]
@thepastaclaw please continue with this PR. Also try to minimize diff size / DRY.

File: src/ui/tools/transition_visualizer_screen.rs [hunk 5]
```diff
@@ -540,3 +581,161 @@ impl ScreenLike for TransitionVisualizerScreen {
...
+        }
+    }
+
+    #[test]
+    fn issue_573_platform_address_transition_deserializes_and_renders() {
+        let bytes = STANDARD
+            .decode(ISSUE_573_BASE64)
+            .expect("issue #573 base64 should decode");
+        let state_transition = StateTransition::deserialize_from_bytes(&bytes)
+            .expect("issue #573 transition should deserialize");
```
Comment (thepastaclaw) [inline src/ui/tools/transition_visualizer_screen.rs]: **🔴 Blocking: Test module does not compile: StateTransition::deserialize_from_bytes no longer exists at the pinned rev**

Line 670 calls StateTransition::deserialize_from_bytes, but the pinned platform rev split that API into trusted/untrusted variants — production parse_input at line 168 already uses deserialize_from_bytes_untrusted, and the file imports PlatformDeserializableUntrusted. CI logs attached to the review thread confirm E0599 on both the Tests and Clippy gates, so the whole lib test target is red. Using deserialize_from_bytes_untrusted also makes the fixture exercise the same untrusted decoding path as production.

```suggestion
        let state_transition = StateTransition::deserialize_from_bytes_untrusted(&bytes)
            .expect("issue #573 transition should deserialize");
```

<sub>source: `glm-5.3-flash` (phase1-reviewer: general)</sub>

File: src/ui/tools/transition_visualizer_screen.rs [hunk 6]
```diff
@@ -540,3 +581,161 @@ impl ScreenLike for TransitionVisualizerScreen {
...
+            !parsed.rendered.trim().is_empty(),
+            "rendered pane must show the parsed transition"
+        );
+        assert!(
+            renders_address_funds_transfer(&parsed.rendered),
+            "rendered pane should identify the AddressFundsTransfer variant"
+        );
+        // Typed transition is retained — broadcast must not need to re-parse rendered text.
+        // Touch the field so a future refactor that drops it fails this test.
+        let _retained: &StateTransition = &parsed.state_transition;
```
Comment (thepastaclaw) [inline src/ui/tools/transition_visualizer_screen.rs]: **🟡 Suggestion: Retention assertion in parse_input_retains_typed_transition_for_issue_573_payload is a compile-time no-op**

The final line let _retained: &StateTransition = &parsed.state_transition; only re-checks the field's type, which surrounding field accesses already force at compile time — it asserts nothing about the value. A refactor storing a default/empty transition would pass trivially, and dropping the field would break compilation regardless of this line. The broadcast fix's core contract (the retained value is exactly the parsed payload) deserves a value assertion such as matching the AddressFundsTransfer variant and round-tripping through serialize_to_bytes back to the original input bytes.

```suggestion
        // Typed transition is retained — broadcast must not need to re-parse rendered text.
        assert!(
            matches!(parsed.state_transition, StateTransition::AddressFundsTransfer(_)),
            "retained transition must be the AddressFundsTransfer variant"
        );
```

<sub>source: `glm-5.3-flash` (phase1-reviewer: general, rust-quality)</sub>

**Review by thepastaclaw — COMMENTED** [review]
## ⚠️ DEGRADED — Re-review — Preliminary review — Phase 1 blocker gate

> **⚠️ DEGRADED review.** The primary review models were unavailable (`gpt-6-astra` unavailable: All credentials for model gpt-6-astra are cooling down (last error: usage_limit_reached: The usage limit has been reache), so this review ran on stand-in models: `gpt-5.6-luna` → `muse-spark-1.3-contributor`, `gpt-5.6-sol` → `muse-spark-1.3-contributor`, `gpt-5.6-terra` → `muse-spark-1.3-contributor`, `gpt-6-astra` → `muse-spark-1.3-contributor`. Both review phases and the independent verifiers still ran, but on weaker models, with Phase 1 capped at `high` effort. Treat the verdict as provisional; a full-strength re-review will run on the next push once the primary models are back.

The production fix is sound — retaining the typed StateTransition decouples broadcast from rendered text and the debug fallback keeps unserializable transitions visible. However the new test calls the removed StateTransition::deserialize_from_bytes, so the lib test target fails to compile and CI gates cannot pass. Two carried-forward test-quality suggestions also still hold at this head.

Validated blockers were found by the Phase-1 review and confirmed by a fresh verifier. Phase 2 is deferred until a fresh same-head revalidation clears the blocker gate.

🔴 1 blocking | 🟡 3 suggestion(s)

_2 carried-forward finding(s) already raised on this PR; not re-posting as new inline comments._

<details>
<summary>Review provenance</summary>

Source: reviewer 1: `glm-5.3-flash` (agent: `phase1-reviewer`, role: `general`); reviewer 2: `glm-5.3-flash` (agent: `phase1-reviewer`, role: `rust-quality`); final verifier: `muse-spark-1.3-contributor` (standing in for `gpt-6-astra`) (agent: `astra-gate-verifier`, role: `verifier`)

- **Degraded mode**: `gpt-6-astra` unavailable: All credentials for model gpt-6-astra are cooling down (last error: usage_limit_reached: The usage limit has been reache (detected by probe, since 2026-09-24T12:50:20Z); stand-ins `gpt-5.6-luna` → `muse-spark-1.3-contributor`, `gpt-5.6-sol` → `muse-spark-1.3-contributor`, `gpt-5.6-terra` → `muse-spark-1.3-contributor`, `gpt-6-astra` → `muse-spark-1.3-contributor`; Phase 1 effort capped at `high`
- Triage: `normal` by `muse-spark-1.3-contributor` (standing in for `gpt-6-astra`) (effort low) — Single-screen UI fix adding a debug fallback render and retaining the parsed transition for broadcast is an ordinary tooling/UI logic change with no critical-surface modification.
- Phase 1 reviewers: `glm-5.3-flash` — general (completed, effort high); agent `phase1-reviewer`, `glm-5.3-flash` — rust-quality (completed, effort high); agent `phase1-reviewer`
- Phase 1 model: `glm-5.3-flash` — zai quota: 5h 95% left, weekly 42% left; passed over `gemini-3.8-flash-high` (antigravity below 15% reserve: weekly 1% left, 5h 100% left)
- Fresh verifier: `muse-spark-1.3-contributor` (standing in for `gpt-6-astra`) — verifier; agent `astra-gate-verifier`
- Phase 2 reviewers: **not run (deferred by blocker gate)**

</details>

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `src/ui/tools/transition_visualizer_screen.rs`:
- [BLOCKING] src/ui/tools/transition_visualizer_screen.rs:670-671: Test module does not compile: StateTransition::deserialize_from_bytes no longer exists at the pinned rev
  Line 670 calls StateTransition::deserialize_from_bytes, but the pinned platform rev split that API into trusted/untrusted variants — production parse_input at line 168 already uses deserialize_from_bytes_untrusted, and the file imports PlatformDeserializableUntrusted. CI logs attached to the review thread confirm E0599 on both the Tests and Clippy gates, so the whole lib test target is red. Using deserialize_from_bytes_untrusted also makes the fixture exercise the same untrusted decoding path as production.
- [SUGGESTION] src/ui/tools/transition_visualizer_screen.rs:737-739: Retention assertion in parse_input_retains_typed_transition_for_issue_573_payload is a compile-time no-op
  The final line let _retained: &StateTransition = &parsed.state_transition; only re-checks the field's type, which surrounding field accesses already force at compile time — it asserts nothing about the value. A refactor storing a default/empty transition would pass trivially, and dropping the field would break compilation regardless of this line. The broadcast fix's core contract (the retained value is exactly the parsed payload) deserves a value assertion such as matching the AddressFundsTransfer variant and round-tripping through serialize_to_bytes back to the original input bytes.
- [SUGGESTION] src/ui/tools/transition_visualizer_screen.rs:684-710: Exercise the serialization-failure fallback deterministically
  (existing thread: https://github.com/dashpay/dash-evo-tool/pull/853#discussion_r3962069718)
  The test probes serde_json::to_string_pretty(&state_transition).is_ok() at runtime and asserts whichever branch runs, so on the current pin — where the AddressFundsTransfer address-map serde helpers reshape maps into entry arrays and serialization succeeds — the Err arm of state_transition_to_display (lines 40-63, the actual code this PR adds for issue #573) is executed by zero tests and a fallback regression can never fail the suite. Because the PR's stated goal includes the debug fallback plus coverage for the #573 payload, the fallback branch deserves a deterministic unit test (e.g. a stub whose Serialize impl always fails) rather than a runtime branch that silently never runs the fallback.
- [SUGGESTION] src/ui/tools/transition_visualizer_screen.rs:604-624: Remove redundant process-global setup from the test fixture
  (existing thread: https://github.com/dashpay/dash-evo-tool/pull/853#discussion_r3962069714)
  ensure_test_env() mutates process-global state the per-test fixture already makes unnecessary: copy_env_file_if_not_exists() writes into the real user data directory outside the test tempdir, and ten unsafe set_var calls inside a Once leak into every other test in the parallel lib-test binary — while ensure_env_file(&data_dir) already seeds each test's own tempdir with the full .env content that Config::load_from reads. The shared crate::context::test_support::test_app_context_for_network fixture already does what this ~55-line fixture needs and would shrink the diff, matching the maintainer's minimize-diff/DRY request.
```
</details>

_Canonical verifier result: `REQUEST_CHANGES`. GitHub does not allow authors to approve or request changes on their own pull requests, so this review was submitted using `COMMENT` transport. The findings and blocker status above are unchanged._
